File: backend/data-transformer/src/fvtt/heritage.ts
import { Exclude, Expose, Transform } from "class-transformer";
import type { Publication } from "../models/secondary/publication";
import { mapToRule, type RuleType } from "../models/secondary/rule";

export class FvttHeritage {
    @Expose({ name: "_id" }) fvttId!: string;

    name!: string;

    @Transform(({ obj }) => obj.system.slug)
    @Expose()
    slug!: string;

    @Transform(({ obj }) =>
        obj.system.ancestry ? obj.system.ancestry.slug : "",
    )
    @Expose()
    ancestrySlug!: string;

    @Transform(({ obj }) => obj.system.publication)
    @Expose()
    publication!: Publication;

    @Transform(({ obj }) => obj.system.rules.map((v: any) => mapToRule(v)))
    @Expose()
    rules!: RuleType[];

    @Transform(({ obj }) => obj.system.traits.rarity)
    @Expose()
    rarity!: string;

    @Transform(({ obj }) => obj.system.traits.value)
    @Expose()
    traits!: string[];

    @Exclude() description!: null;
    @Exclude() folder!: null;
    @Exclude() _stats!: null;
    @Exclude() type!: null;
    @Exclude() img!: null;
    @Exclude() system!: null;
}

File: backend/src/models/primary/heritage.rs
use crate::helpers::null_to_default;
use crate::models::{Publication, Rule, Summary, SummaryData};
use crate::storeable::Storeable;
use bson::oid::ObjectId;
use serde::{Deserialize, Serialize};
use specta::Type;

#[derive(Serialize, Deserialize, Debug, Type)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Heritage {
    #[serde(rename = "_id")]
    pub mongo_id: ObjectId,
    pub fvtt_id: String,
    pub name: String,
    pub slug: String,
    pub description: String,
    pub ancestry_slug: String,
    pub publication: Publication,
    #[serde(default, deserialize_with = "null_to_default")]
    pub rules: Vec<Rule>,
    pub rarity: String,
    #[serde(default, deserialize_with = "null_to_default")]
    pub traits: Vec<String>,
    #[serde(rename = "type")]
    pub data_type: String,
}

impl Storeable for Heritage {
    fn table_name() -> &'static str {
        "heritage"
    }
}

impl From<Heritage> for Summary {
    fn from(value: Heritage) -> Self {
        let data = vec![SummaryData::String {
            value: value.name.clone(),
            link: None,
            tooltip: None,
            abbreviation: None,
        }];

        Self {
            id: value.fvtt_id,
            name: value.name,
            slug: value.slug,
            data,
        }
    }
}
